src/halmos/sevm.py

```diff
@@ -1894,192 +1925,69 @@ def mk_mod(self, ex: Exec, x: Any, y: Any) -> Any:
 
         if op == EVM.SUB:
-            return w1 - w2
+            return w1.sub(w2)
 
         if op == EVM.MUL:
-            is_bv_value_w1 = is_bv_value(w1)
-            is_bv_value_w2 = is_bv_value(w2)
-
-            if is_bv_value_w1 and is_bv_value_w2:
-                return w1 * w2
-
-            if is_bv_value_w1:
-                i1: int = w1.as_long()
-                if i1 == 0:
-                    return w1
-
-                if i1 == 1:
-                    return w2
-
-                if is_power_of_two(i1):
-                    return w2 << (i1.bit_length() - 1)
-
-            if is_bv_value_w2:
-                i2: int = w2.as_long()
-                if i2 == 0:
-                    return w2
-
-                if i2 == 1:
-                    return w1
-
-                if is_power_of_two(i2):
-                    return w1 << (i2.bit_length() - 1)
-
-            if is_bv_value_w1 or is_bv_value_w2:
-                return w1 * w2
-
-            return self.mk_mul(ex, w1, w2)
+            return w1.mul(w2, abstraction=f_mul[w1.size])
 
         if op == EVM.DIV:
-            div_for_overflow_check = self.div_xy_y(w1, w2)
-            if div_for_overflow_check is not None:  # xy/x or xy/y
-                return div_for_overflow_check
-
-            if is_bv_value(w1) and is_bv_value(w2):
-                if w2.as_long() == 0:
-                    return w2
-                else:
-                    return UDiv(w1, w2)  # unsigned div (bvudiv)
-
-            if is_bv_value(w2):
-                # concrete denominator case
-                i2: int = w2.as_long()
-                if i2 == 0:
-                    return w2
-
-                if i2 == 1:
-                    return w1
+            # TODO: div_xy_y
 
-                if is_power_of_two(i2):
-                    return LShR(w1, i2.bit_length() - 1)
-
-            return self.mk_div(ex, w1, w2)
+            term = w1.div(w2, abstraction=f_div)
+            if term.is_symbolic:
+                ex.path.append(ULE(term.as_z3(), w1.as_z3()))  # (x / y) <= x
+            return term
 
         if op == EVM.MOD:
-            if is_bv_value(w1) and is_bv_value(w2):
-                if w2.as_long() == 0:
-                    return w2
-                else:
-                    return URem(w1, w2)  # bvurem
-
-            if is_bv_value(w2):
-                i2: int = int(str(w2))
-                if i2 == 0 or i2 == 1:
-                    return con(0, w2.size())
-
-                if is_power_of_two(i2):
-                    bitsize = i2.bit_length() - 1
-                    return ZeroExt(w2.size() - bitsize, Extract(bitsize - 1, 0, w1))
-
-            return self.mk_mod(ex, w1, w2)
+            term = w1.mod(w2, abstraction=f_mod[w1.size])
+            if term.is_symbolic:
+                # (x % y) <= y
+                # not ULT, because y could be 0 and x % 0 = 0
+                ex.path.append(ULE(term.as_z3(), w2.as_z3()))
+            return term
```

Comment:
not sure why yet but this seems to cause a problem in `check_Mod`, even though we do add `ULE(f_evm_bvurem_256(p_x_uint256_9834528_00, p_y_uint256_7a9fba7_00), p_y_uint256_7a9fba7_00)` to the path

this is the failing assert:

```
uint x_mod_y = unchecked_mod(x, y);
assert(x_mod_y <= y);
```

🤨